src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs

```diff
@@ -73,77 +75,206 @@ public string StartPreparingPayload(BlockHeader parentHeader, PayloadAttributes
         blockImprovementContext.ImprovementTask.ContinueWith(
-            (b) => LogProductionResult(b, currentBestBlock, blockImprovementContext.BlockFees, Stopwatch.GetElapsedTime(startTimestamp)),
+            (b) =>
+            {
+                if (!cts.IsCancellationRequested)
+                {
+                    LogProductionResult(b, currentBestBlock, blockImprovementContext.BlockFees, Stopwatch.GetElapsedTime(startTimestamp));
+                }
+            },
             TaskContinuationOptions.RunContinuationsAsynchronously);
         blockImprovementContext.ImprovementTask.ContinueWith(async _ =>
         {
-            // if after delay we still have time to try producing the block in this slot
+            CancellationToken token = cts.Token;
+            if (token.IsCancellationRequested)
+            {
+                return;
+            }
+
+            TimeSpan dynamicDelay = CalculateImprovementDelay(startDateTime);
             TimeSpan lastBuildTime = Stopwatch.GetElapsedTime(startTimestamp);
-            DateTimeOffset whenWeCouldFinishNextProduction = DateTimeOffset.UtcNow + _improvementDelay + lastBuildTime;
-            DateTimeOffset slotFinished = startDateTime + _timePerSlot;
-            if (whenWeCouldFinishNextProduction < slotFinished)
+            DateTimeOffset whenWeCouldFinishNextProduction = DateTimeOffset.UtcNow + dynamicDelay + lastBuildTime;
+            // We don't want to keep improving a block too far beyond the slot duration.
+            // Specifically, we allow ourselves at most 30% extra of the nominal slot time.
+            // This is just a break in case the improvement is never cancelled.
+            DateTimeOffset slotPlusThirdFinishTime = startDateTime + _timePerSlot * 1.3;
+            if (whenWeCouldFinishNextProduction > slotPlusThirdFinishTime)
             {
-                if (_logger.IsTrace) _logger.Trace($"Block for payload {payloadId} with parent {parentHeader.ToString(BlockHeader.Format.FullHashAndNumber)} will be improved in {_improvementDelay.TotalMilliseconds}ms");
-                await Task.Delay(_improvementDelay);
-                if (!blockImprovementContext.Disposed) // if GetPayload wasn't called for this item or it wasn't cleared
-                {
-                    Block newBestBlock = blockImprovementContext.CurrentBestBlock ?? currentBestBlock;
-                    ImproveBlock(payloadId, parentHeader, payloadAttributes, newBestBlock, startDateTime, blockImprovementContext.BlockFees);
-                }
-                else
+                // If we can't finish before that cutoff, skip the improvement
+                if (_logger.IsTrace) _logger.Trace($"Block for payload {payloadId} with parent {parentHeader.ToString(BlockHeader.Format.FullHashAndNumber)} won't be improved, no more time in slot");
+                return;
+            }
```

Comment:
I would put this part into separate method too

Reply:
Moved